components/scheduler2/generator_executors.go

```diff
@@ -0,0 +1,163 @@
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	GeneratorTaskExecutorOptions struct {
+		fx.In
+
+		Config         *Config
+		MetricsHandler metrics.Handler
+		Logger         log.Logger
+		SpecProcessor  *SpecProcessor
+	}
+
+	generatorTaskExecutor struct {
+		GeneratorTaskExecutorOptions
+	}
+)
+
+func RegisterGeneratorExecutors(registry *hsm.Registry, options GeneratorTaskExecutorOptions) error {
+	e := generatorTaskExecutor{options}
+	if err := hsm.RegisterTimerExecutor(registry, e.executeBufferTask); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e generatorTaskExecutor) executeBufferTask(env hsm.Environment, node *hsm.Node, task BufferTask) error {
+	schedulerNode := node.Parent
+	scheduler, err := e.loadScheduler(schedulerNode)
+	if err != nil {
+		return err
+	}
+
+	generator, err := e.loadGenerator(node)
+	if err != nil {
+		return err
+	}
+
+	// If we have no last processed time, this is a new schedule.
+	if generator.LastProcessedTime == nil {
+		generator.LastProcessedTime = timestamppb.Now()
+		// TODO - update schedule info with create time
+
+		e.logSchedule("Starting schedule", scheduler)
+	}
+
+	// Process time range between last high water mark and system time.
+	t1 := generator.LastProcessedTime.AsTime()
+	t2 := time.Now().UTC()
+	if t2.Before(t1) {
+		e.Logger.Warn("Time went backwards",
+			tag.NewStringTag("time", t1.String()),
+			tag.NewStringTag("time", t2.String()))
+		t2 = t1
+	}
+
+	res, err := e.SpecProcessor.ProcessTimeRange(scheduler, t1, t2, false, nil)
+	if err != nil {
+		// An error here should be impossible, send to the DLQ.
+		e.Logger.Error("Error processing time range", tag.Error(err))
+
+		return fmt.Errorf(
+			"%w: %w",
+			err,
+			serviceerror.NewInternal("Scheduler's Generator failed to process a time range"),
+		)
+	}
+
+	// Transition the executor sub state machine to execute the new buffered actions.
+	executorNode, err := schedulerNode.Child([]hsm.Key{ExecutorMachineKey})
+	if err != nil {
+		return fmt.Errorf(
+			"%w: %w",
+			err,
+			serviceerror.NewInternal("Scheduler is missing its Executor node"),
+		)
+	}
+	err = hsm.MachineTransition(executorNode, func(e Executor) (hsm.TransitionOutput, error) {
+		return TransitionExecute.Apply(e, EventExecute{
+			Node:            executorNode,
+			BufferedActions: res.BufferedStarts,
+		})
+	})
+	if err != nil {
+		return fmt.Errorf(
+			"%w: unable to transition Executor to Executing state",
+			err,
+		)
+	}
+
+	// Write Generator internal state, flushing the high water mark to persistence.
+	// Another buffering task is also added. if we fail here (after transitioning
+	// Executor), that's fine; buffered actions are idempotent through their generated
+	// request IDs.
+	err = hsm.MachineTransition(node, func(g Generator) (hsm.TransitionOutput, error) {
+		wakeupTime := res.NextWakeupTime
+		g.LastProcessedTime = timestamppb.New(res.LastActionTime)
+		g.NextInvocationTime = timestamppb.New(wakeupTime)
```

Comment:
I'd be careful caching this property, it may be better to compute it on the fly and store it only on the generated task's deadline.